File: libs/linux-abi/probe/virtgpu.c
/*
 * The virtio-gpu numbers and layouts `libs/linux-abi/src/virtgpu.rs` writes
 * down, printed from the UAPI headers themselves.
 *
 * Run by `libs/linux-abi/probe/virtgpu.sh` on a Linux host with the kernel's
 * UAPI headers (`linux-libc-dev`) once natively for 64-bit and once for
 * ARMv7-A under qemu-arm, into `virtgpu-64.txt` and `virtgpu-32.txt` beside
 * it. The crate's tests read both files and require every line to match what
 * the crate says, so a number changed on either side fails `cargo test`.
 *
 * Each line is `name value`, in decimal. Layouts are `sizeof.<struct>` and
 * `offsetof.<struct>.<field>`.
 *
 * These are the render node's ioctls, which `/dev/dri/renderD<N>` answers and
 * `/dev/dri/card0` does not, so they are probed apart from `drm.c`: they come
 * from one driver's header rather than from DRM's own, and every one of them
 * is the same at both widths, because the structures carry their user
 * pointers as `__u64`.
 */

#include <stddef.h>
#include <stdio.h>
#include <drm/drm.h>
#include <drm/virtgpu_drm.h>

#define VALUE(name) printf("%s %llu\n", #name, (unsigned long long)(name))
#define SIZE(type) printf("sizeof.%s %zu\n", #type, sizeof(struct type))
#define FIELD(type, field) \
	printf("offsetof.%s.%s %zu\n", #type, #field, offsetof(struct type, field))

int main(void)
{
	/*
	 * The ioctls. Driver-private, so each number is `DRM_COMMAND_BASE`
	 * plus the driver's own index, and encodes its argument's size.
	 */
	VALUE(DRM_IOCTL_VIRTGPU_MAP);
	VALUE(DRM_IOCTL_VIRTGPU_EXECBUFFER);
	VALUE(DRM_IOCTL_VIRTGPU_GETPARAM);
	VALUE(DRM_IOCTL_VIRTGPU_RESOURCE_CREATE);
	VALUE(DRM_IOCTL_VIRTGPU_RESOURCE_INFO);
	VALUE(DRM_IOCTL_VIRTGPU_TRANSFER_FROM_HOST);
	VALUE(DRM_IOCTL_VIRTGPU_TRANSFER_TO_HOST);
	VALUE(DRM_IOCTL_VIRTGPU_WAIT);
	VALUE(DRM_IOCTL_VIRTGPU_GET_CAPS);
	VALUE(DRM_IOCTL_VIRTGPU_RESOURCE_CREATE_BLOB);
	VALUE(DRM_IOCTL_VIRTGPU_CONTEXT_INIT);

	/* What GETPARAM answers about. */
	VALUE(VIRTGPU_PARAM_3D_FEATURES);
	VALUE(VIRTGPU_PARAM_CAPSET_QUERY_FIX);
	VALUE(VIRTGPU_PARAM_RESOURCE_BLOB);
	VALUE(VIRTGPU_PARAM_HOST_VISIBLE);
	VALUE(VIRTGPU_PARAM_CROSS_DEVICE);
	VALUE(VIRTGPU_PARAM_CONTEXT_INIT);
	VALUE(VIRTGPU_PARAM_SUPPORTED_CAPSET_IDs);
	VALUE(VIRTGPU_PARAM_EXPLICIT_DEBUG_NAME);

	/* The capability sets GET_CAPS may be asked for. */
	VALUE(VIRTGPU_DRM_CAPSET_VIRGL);
	VALUE(VIRTGPU_DRM_CAPSET_VIRGL2);
	VALUE(VIRTGPU_DRM_CAPSET_VENUS);

	SIZE(drm_virtgpu_getparam);
	FIELD(drm_virtgpu_getparam, param);
	FIELD(drm_virtgpu_getparam, value);

	SIZE(drm_virtgpu_context_init);
	FIELD(drm_virtgpu_context_init, num_params);
	FIELD(drm_virtgpu_context_init, pad);
	FIELD(drm_virtgpu_context_init, ctx_set_params);

	SIZE(drm_virtgpu_context_set_param);
	FIELD(drm_virtgpu_context_set_param, param);
	FIELD(drm_virtgpu_context_set_param, value);

	/* What CONTEXT_INIT's parameters may set. */
	VALUE(VIRTGPU_CONTEXT_PARAM_CAPSET_ID);
	VALUE(VIRTGPU_CONTEXT_PARAM_NUM_RINGS);
	VALUE(VIRTGPU_CONTEXT_PARAM_POLL_RINGS_MASK);
	VALUE(VIRTGPU_CONTEXT_PARAM_DEBUG_NAME);

	SIZE(drm_virtgpu_resource_create);
	FIELD(drm_virtgpu_resource_create, target);
	FIELD(drm_virtgpu_resource_create, format);
	FIELD(drm_virtgpu_resource_create, bind);
	FIELD(drm_virtgpu_resource_create, width);
	FIELD(drm_virtgpu_resource_create, height);
	FIELD(drm_virtgpu_resource_create, depth);
	FIELD(drm_virtgpu_resource_create, array_size);
	FIELD(drm_virtgpu_resource_create, last_level);
	FIELD(drm_virtgpu_resource_create, nr_samples);
	FIELD(drm_virtgpu_resource_create, flags);
	FIELD(drm_virtgpu_resource_create, bo_handle);
	FIELD(drm_virtgpu_resource_create, res_handle);
	FIELD(drm_virtgpu_resource_create, size);
	FIELD(drm_virtgpu_resource_create, stride);

	SIZE(drm_virtgpu_resource_create_blob);
	FIELD(drm_virtgpu_resource_create_blob, blob_mem);
	FIELD(drm_virtgpu_resource_create_blob, blob_flags);
	FIELD(drm_virtgpu_resource_create_blob, bo_handle);
	FIELD(drm_virtgpu_resource_create_blob, res_handle);
	FIELD(drm_virtgpu_resource_create_blob, size);
	FIELD(drm_virtgpu_resource_create_blob, pad);
	FIELD(drm_virtgpu_resource_create_blob, cmd_size);
	FIELD(drm_virtgpu_resource_create_blob, cmd);
	FIELD(drm_virtgpu_resource_create_blob, blob_id);

	/* Where a blob's memory is, and what it may be used for. */
	VALUE(VIRTGPU_BLOB_MEM_GUEST);
	VALUE(VIRTGPU_BLOB_MEM_HOST3D);
	VALUE(VIRTGPU_BLOB_MEM_HOST3D_GUEST);
	VALUE(VIRTGPU_BLOB_FLAG_USE_MAPPABLE);
	VALUE(VIRTGPU_BLOB_FLAG_USE_SHAREABLE);
	VALUE(VIRTGPU_BLOB_FLAG_USE_CROSS_DEVICE);

	SIZE(drm_virtgpu_resource_info);
	FIELD(drm_virtgpu_resource_info, bo_handle);
	FIELD(drm_virtgpu_resource_info, res_handle);
	FIELD(drm_virtgpu_resource_info, size);
	FIELD(drm_virtgpu_resource_info, blob_mem);

	SIZE(drm_virtgpu_map);
	FIELD(drm_virtgpu_map, offset);
	FIELD(drm_virtgpu_map, handle);
	FIELD(drm_virtgpu_map, pad);

	SIZE(drm_virtgpu_get_caps);
	FIELD(drm_virtgpu_get_caps, cap_set_id);
	FIELD(drm_virtgpu_get_caps, cap_set_ver);
	FIELD(drm_virtgpu_get_caps, addr);
	FIELD(drm_virtgpu_get_caps, size);
	FIELD(drm_virtgpu_get_caps, pad);

	SIZE(drm_virtgpu_execbuffer);
	FIELD(drm_virtgpu_execbuffer, flags);
	FIELD(drm_virtgpu_execbuffer, size);
	FIELD(drm_virtgpu_execbuffer, command);
	FIELD(drm_virtgpu_execbuffer, bo_handles);
	FIELD(drm_virtgpu_execbuffer, num_bo_handles);
	FIELD(drm_virtgpu_execbuffer, fence_fd);
	FIELD(drm_virtgpu_execbuffer, ring_idx);
	FIELD(drm_virtgpu_execbuffer, syncobj_stride);
	FIELD(drm_virtgpu_execbuffer, num_in_syncobjs);
	FIELD(drm_virtgpu_execbuffer, num_out_syncobjs);
	FIELD(drm_virtgpu_execbuffer, in_syncobjs);
	FIELD(drm_virtgpu_execbuffer, out_syncobjs);

	/* What a submission may ask for, and which ring it runs on. */
	VALUE(VIRTGPU_EXECBUF_FENCE_FD_IN);
	VALUE(VIRTGPU_EXECBUF_FENCE_FD_OUT);
	VALUE(VIRTGPU_EXECBUF_RING_IDX);
	/* What a wait may ask for. */
	VALUE(VIRTGPU_WAIT_NOWAIT);

	SIZE(drm_virtgpu_3d_box);
	FIELD(drm_virtgpu_3d_box, x);
	FIELD(drm_virtgpu_3d_box, y);
	FIELD(drm_virtgpu_3d_box, z);
	FIELD(drm_virtgpu_3d_box, w);
	FIELD(drm_virtgpu_3d_box, h);
	FIELD(drm_virtgpu_3d_box, d);

	SIZE(drm_virtgpu_3d_transfer_to_host);
	FIELD(drm_virtgpu_3d_transfer_to_host, bo_handle);
	FIELD(drm_virtgpu_3d_transfer_to_host, box);
	FIELD(drm_virtgpu_3d_transfer_to_host, level);
	FIELD(drm_virtgpu_3d_transfer_to_host, offset);
	FIELD(drm_virtgpu_3d_transfer_to_host, stride);
	FIELD(drm_virtgpu_3d_transfer_to_host, layer_stride);

	SIZE(drm_virtgpu_3d_transfer_from_host);
	FIELD(drm_virtgpu_3d_transfer_from_host, bo_handle);
	FIELD(drm_virtgpu_3d_transfer_from_host, box);
	FIELD(drm_virtgpu_3d_transfer_from_host, level);
	FIELD(drm_virtgpu_3d_transfer_from_host, offset);
	FIELD(drm_virtgpu_3d_transfer_from_host, stride);
	FIELD(drm_virtgpu_3d_transfer_from_host, layer_stride);

	SIZE(drm_virtgpu_3d_wait);
	FIELD(drm_virtgpu_3d_wait, handle);
	FIELD(drm_virtgpu_3d_wait, flags);
	return 0;
}

File: libs/linux-abi/src/virtgpu.rs
//! virtio-gpu: the ioctls, constants and structures a 3D client uses on
//! `/dev/dri/renderD<N>`.
//!
//! This is the render node's side of `docs/GPU.md`'s seam. [`crate::drm`] is
//! the display: modes, framebuffers and page flips on `/dev/dri/card0`, which
//! every DRM driver answers the same way. These are one driver's own ioctls,
//! numbered from `DRM_COMMAND_BASE`, and a card never answers them.
//!
//! # Where the numbers come from
//!
//! `include/uapi/drm/virtgpu_drm.h`, which every architecture takes
//! unchanged. `probe/virtgpu.c` prints every number and layout below from
//! that header, natively for 64-bit and under `qemu-arm` for ARMv7-A, into
//! `probe/virtgpu-64.txt` and `probe/virtgpu-32.txt`; the tests read both
//! files and require this module to agree with every line.
//!
//! # No structure here has a width
//!
//! Unlike [`crate::drm::Version`], which holds `size_t` lengths and `char *`
//! pointers and so differs between the widths, every structure below carries
//! its user pointers as `__u64` and its sizes as `__u32`. All of them are the
//! same size with the same offsets on x86-64 and on ARMv7-A, which is why
//! none of them takes a [`crate::socket::Width`], and the tests assert that
//! the two probe files are identical rather than listing exceptions.
//!
//! # Reading and writing
//!
//! Every structure is read from and written into the bytes the ioctl's
//! argument points at, little-endian, returning `None` rather than panicking
//! when the buffer is short. What the fields mean is the device's business;
//! nothing here validates a value.

use crate::layout::layout;
pub use crate::layout::{Field, Layout};

// ---------------------------------------------------------------------------
// ioctls
// ---------------------------------------------------------------------------

/// `DRM_IOCTL_VIRTGPU_MAP`: ask for the offset to `mmap` an object at.
pub const IOCTL_MAP: u32 = 0xC010_6441;
/// `DRM_IOCTL_VIRTGPU_EXECBUFFER`: submit a command buffer to a context.
pub const IOCTL_EXECBUFFER: u32 = 0xC040_6442;
/// `DRM_IOCTL_VIRTGPU_GETPARAM`: ask what the device supports.
pub const IOCTL_GETPARAM: u32 = 0xC010_6443;
/// `DRM_IOCTL_VIRTGPU_RESOURCE_CREATE`: create a 3D resource and its object.
pub const IOCTL_RESOURCE_CREATE: u32 = 0xC038_6444;
/// `DRM_IOCTL_VIRTGPU_RESOURCE_INFO`: the resource behind an object.
pub const IOCTL_RESOURCE_INFO: u32 = 0xC010_6445;
/// `DRM_IOCTL_VIRTGPU_TRANSFER_FROM_HOST`: read a resource back.
pub const IOCTL_TRANSFER_FROM_HOST: u32 = 0xC02C_6446;
/// `DRM_IOCTL_VIRTGPU_TRANSFER_TO_HOST`: send a resource's contents over.
pub const IOCTL_TRANSFER_TO_HOST: u32 = 0xC02C_6447;
/// `DRM_IOCTL_VIRTGPU_WAIT`: wait until an object is idle.
pub const IOCTL_WAIT: u32 = 0xC008_6448;
/// `DRM_IOCTL_VIRTGPU_GET_CAPS`: read a capability set.
pub const IOCTL_GET_CAPS: u32 = 0xC018_6449;
/// `DRM_IOCTL_VIRTGPU_RESOURCE_CREATE_BLOB`: create a blob resource. The
/// number is written down so the node can tell it apart from a call it has
/// not heard of; blob resources are later.
pub const IOCTL_RESOURCE_CREATE_BLOB: u32 = 0xC030_644A;
/// `DRM_IOCTL_VIRTGPU_CONTEXT_INIT`: set a context's parameters.
pub const IOCTL_CONTEXT_INIT: u32 = 0xC010_644B;

// ---------------------------------------------------------------------------
// What GETPARAM answers about
// ---------------------------------------------------------------------------

/// `VIRTGPU_PARAM_3D_FEATURES`: the device does 3D at all.
pub const PARAM_3D_FEATURES: u64 = 1;
/// `VIRTGPU_PARAM_CAPSET_QUERY_FIX`.
pub const PARAM_CAPSET_QUERY_FIX: u64 = 2;
/// `VIRTGPU_PARAM_RESOURCE_BLOB`.
pub const PARAM_RESOURCE_BLOB: u64 = 3;
/// `VIRTGPU_PARAM_HOST_VISIBLE`.
pub const PARAM_HOST_VISIBLE: u64 = 4;
/// `VIRTGPU_PARAM_CROSS_DEVICE`.
pub const PARAM_CROSS_DEVICE: u64 = 5;
/// `VIRTGPU_PARAM_CONTEXT_INIT`: contexts take parameters.
pub const PARAM_CONTEXT_INIT: u64 = 6;
/// `VIRTGPU_PARAM_SUPPORTED_CAPSET_IDs`, spelled with that lowercase `s` in
/// the header.
pub const PARAM_SUPPORTED_CAPSET_IDS: u64 = 7;
/// `VIRTGPU_PARAM_EXPLICIT_DEBUG_NAME`.
pub const PARAM_EXPLICIT_DEBUG_NAME: u64 = 8;

// ---------------------------------------------------------------------------
// Capability sets
// ---------------------------------------------------------------------------

/// `VIRTGPU_DRM_CAPSET_VIRGL`: virgl's first capability set.
pub const CAPSET_VIRGL: u32 = 1;
/// `VIRTGPU_DRM_CAPSET_VIRGL2`: virgl's second.
pub const CAPSET_VIRGL2: u32 = 2;

// ---------------------------------------------------------------------------
// Context parameters
// ---------------------------------------------------------------------------

/// `VIRTGPU_CONTEXT_PARAM_CAPSET_ID`: which capability set the context is.
pub const CONTEXT_PARAM_CAPSET_ID: u64 = 1;
/// `VIRTGPU_CONTEXT_PARAM_NUM_RINGS`: how many command rings it has.
pub const CONTEXT_PARAM_NUM_RINGS: u64 = 2;
/// `VIRTGPU_CONTEXT_PARAM_POLL_RINGS_MASK`: which rings a poll reports.
pub const CONTEXT_PARAM_POLL_RINGS_MASK: u64 = 3;
/// `VIRTGPU_CONTEXT_PARAM_DEBUG_NAME`: a name for the host's logs.
pub const CONTEXT_PARAM_DEBUG_NAME: u64 = 4;

// ---------------------------------------------------------------------------
// What a submission asks for
// ---------------------------------------------------------------------------

/// `VIRTGPU_WAIT_NOWAIT`: answer whether the object is idle rather than
/// wait for it, `EBUSY` for not yet.
pub const WAIT_NOWAIT: u32 = 1;

/// `VIRTGPU_EXECBUF_FENCE_FD_IN`: wait on `fence_fd` before running.
pub const EXECBUF_FENCE_FD_IN: u32 = 1;
/// `VIRTGPU_EXECBUF_FENCE_FD_OUT`: return a fence in `fence_fd`.
pub const EXECBUF_FENCE_FD_OUT: u32 = 2;
/// `VIRTGPU_EXECBUF_RING_IDX`: `ring_idx` says which ring to run on.
pub const EXECBUF_RING_IDX: u32 = 4;

// ---------------------------------------------------------------------------
// Layouts
// ---------------------------------------------------------------------------

layout! {
    /// `struct drm_virtgpu_getparam`: `VIRTGPU_GETPARAM`'s argument.
    GetParam = "drm_virtgpu_getparam", 16 {
        /// Which parameter, a `PARAM_*`.
        param: u64 = 0 / "param",
        /// The user address of the `u64` the device writes the answer into.
        value: u64 = 8 / "value",
    }
}

layout! {
    /// `struct drm_virtgpu_context_init`: `VIRTGPU_CONTEXT_INIT`'s argument.
    ContextInit = "drm_virtgpu_context_init", 16 {
        /// How many parameters `ctx_set_params` points at.
        num_params: u32 = 0 / "num_params",
        /// Padding, zero.
        pad: u32 = 4 / "pad",
        /// The user address of that many [`ContextSetParam`].
        ctx_set_params: u64 = 8 / "ctx_set_params",
    }
}

layout! {
    /// `struct drm_virtgpu_context_set_param`: one of [`ContextInit`]'s.
    ContextSetParam = "drm_virtgpu_context_set_param", 16 {
        /// Which parameter, a `CONTEXT_PARAM_*`.
        param: u64 = 0 / "param",
        /// What to set it to.
        value: u64 = 8 / "value",
    }
}

layout! {
    /// `struct drm_virtgpu_resource_create`: `VIRTGPU_RESOURCE_CREATE`'s
    /// argument. The device fills in `bo_handle` and `res_handle`.
    ResourceCreate = "drm_virtgpu_resource_create", 56 {
        /// The virgl target, such as a buffer or a 2D texture.
        target: u32 = 0 / "target",
        /// The virgl format.
        format: u32 = 4 / "format",
        /// What the resource may be bound as.
        bind: u32 = 8 / "bind",
        /// Width in pixels, or bytes for a buffer.
        width: u32 = 12 / "width",
        /// Height in pixels.
        height: u32 = 16 / "height",
        /// Depth.
        depth: u32 = 20 / "depth",
        /// How many array layers.
        array_size: u32 = 24 / "array_size",
        /// The last mip level.
        last_level: u32 = 28 / "last_level",
        /// How many samples.
        nr_samples: u32 = 32 / "nr_samples",
        /// Resource flags.
        flags: u32 = 36 / "flags",
        /// An existing object to attach to, or zero for a new one.
        bo_handle: u32 = 40 / "bo_handle",
        /// The resource, filled in by the device.
        res_handle: u32 = 44 / "res_handle",
        /// How many bytes of memory to back it with.
        size: u32 = 48 / "size",
        /// Bytes per row, which the host validates transfers against.
        stride: u32 = 52 / "stride",
    }
}

layout! {
    /// `struct drm_virtgpu_resource_info`: `VIRTGPU_RESOURCE_INFO`'s
    /// argument.
    ResourceInfo = "drm_virtgpu_resource_info", 16 {
        /// The object asked about.
        bo_handle: u32 = 0 / "bo_handle",
        /// Its resource, filled in by the device.
        res_handle: u32 = 4 / "res_handle",
        /// Its size in bytes.
        size: u32 = 8 / "size",
        /// Which kind of blob memory it is, or zero.
        blob_mem: u32 = 12 / "blob_mem",
    }
}

layout! {
    /// `struct drm_virtgpu_map`: `VIRTGPU_MAP`'s argument. The offset comes
    /// back to be passed to `mmap`, and is not an address.
    Map = "drm_virtgpu_map", 16 {
        /// The offset to `mmap` at, filled in by the device.
        offset: u64 = 0 / "offset",
        /// The object to map.
        handle: u32 = 8 / "handle",
        /// Padding, zero.
        pad: u32 = 12 / "pad",
    }
}

layout! {
    /// `struct drm_virtgpu_get_caps`: `VIRTGPU_GET_CAPS`'s argument.
    GetCaps = "drm_virtgpu_get_caps", 24 {
        /// Which capability set, a `CAPSET_*`.
        cap_set_id: u32 = 0 / "cap_set_id",
        /// Which version of it.
        cap_set_ver: u32 = 4 / "cap_set_ver",
        /// The user address to write it to.
        addr: u64 = 8 / "addr",
        /// How many bytes are there to write into.
        size: u32 = 16 / "size",
        /// Padding, zero.
        pad: u32 = 20 / "pad",
    }
}

layout! {
    /// `struct drm_virtgpu_execbuffer`: `VIRTGPU_EXECBUFFER`'s argument.
    ExecBuffer = "drm_virtgpu_execbuffer", 64 {
        /// `EXECBUF_*`.
        flags: u32 = 0 / "flags",
        /// How many bytes of commands.
        size: u32 = 4 / "size",
        /// The user address of the command buffer.
        command: u64 = 8 / "command",
        /// The user address of an array of object handles.
        bo_handles: u64 = 16 / "bo_handles",
        /// How many handles are in it.
        num_bo_handles: u32 = 24 / "num_bo_handles",
        /// A fence to wait on or return, by `flags`. Signed: `-1` is none.
        fence_fd: i32 = 28 / "fence_fd",
        /// Which command ring, when `EXECBUF_RING_IDX` is set.
        ring_idx: u32 = 32 / "ring_idx",
        /// The size of one sync object entry.
        syncobj_stride: u32 = 36 / "syncobj_stride",
        /// How many sync objects to wait on.
        num_in_syncobjs: u32 = 40 / "num_in_syncobjs",
        /// How many to signal.
        num_out_syncobjs: u32 = 44 / "num_out_syncobjs",
        /// The user address of the ones to wait on.
        in_syncobjs: u64 = 48 / "in_syncobjs",
        /// The user address of the ones to signal.
        out_syncobjs: u64 = 56 / "out_syncobjs",
    }
}

layout! {
    /// `struct drm_virtgpu_3d_box`: the region a transfer covers.
    Box3d = "drm_virtgpu_3d_box", 24 {
        /// Left.
        x: u32 = 0 / "x",
        /// Top.
        y: u32 = 4 / "y",
        /// Front.
        z: u32 = 8 / "z",
        /// Width.
        w: u32 = 12 / "w",
        /// Height.
        h: u32 = 16 / "h",
        /// Depth.
        d: u32 = 20 / "d",
    }
}

layout! {
    /// `struct drm_virtgpu_3d_transfer_to_host`: `VIRTGPU_TRANSFER_TO_HOST`'s
    /// argument.
    TransferToHost = "drm_virtgpu_3d_transfer_to_host", 44 {
        /// The object to send.
        bo_handle: u32 = 0 / "bo_handle",
        /// Which region of it.
        r#box: Box3d = 4 / "box",
        /// Which mip level.
        level: u32 = 28 / "level",
        /// Where in the object to start.
        offset: u32 = 32 / "offset",
        /// Bytes per row.
        stride: u32 = 36 / "stride",
        /// Bytes per layer.
        layer_stride: u32 = 40 / "layer_stride",
    }
}

layout! {
    /// `struct drm_virtgpu_3d_transfer_from_host`:
    /// `VIRTGPU_TRANSFER_FROM_HOST`'s argument.
    TransferFromHost = "drm_virtgpu_3d_transfer_from_host", 44 {
        /// The object to read back into.
        bo_handle: u32 = 0 / "bo_handle",
        /// Which region of it.
        r#box: Box3d = 4 / "box",
        /// Which mip level.
        level: u32 = 28 / "level",
        /// Where in the object to start.
        offset: u32 = 32 / "offset",
        /// Bytes per row.
        stride: u32 = 36 / "stride",
        /// Bytes per layer.
        layer_stride: u32 = 40 / "layer_stride",
    }
}

layout! {
    /// `struct drm_virtgpu_3d_wait`: `VIRTGPU_WAIT`'s argument.
    Wait = "drm_virtgpu_3d_wait", 8 {
        /// The object to wait on. Zero is not a handle.
        handle: u32 = 0 / "handle",
        /// Wait flags.
        flags: u32 = 4 / "flags",
    }
}
